rasa/nlu/extractors/crf_entity_extractor.py

```diff
@@ -572,14 +582,14 @@ def _train_model(self, df_train: List[List[CRFToken]]) -> None:
         for tag_name in self.crf_order:
             logger.debug(f"Training CRF for '{tag_name}'.")
 
-            # add entity tag features for second level CRFs
-            include_tag_features = tag_name != ENTITY_ATTRIBUTE_TYPE
+            # this could be optimized if necessary
             X_train = (
-                self._crf_tokens_to_features(sentence, include_tag_features)
-                for sentence in df_train
+                self._message_to_features(msg, tag_name)
+                for msg in training_data.stream_featurized_messages()
             )
             y_train = (
-                self._crf_tokens_to_tags(sentence, tag_name) for sentence in df_train
+                self._message_to_tags(msg, tag_name)
+                for msg in training_data.stream_featurized_messages()
```

Comment:
that's potentially dangerous if shuffle would be enabled as X and Y would be shuffled differently, right?

Reply:
Yes, indeed that would break. The reason this is here is that you can only consume a generator once. However, you could split the generator, if you can guarantee that no copy is advancing more than `n` elements ahead of the others. It would be possible to do this here as the crf internally zips x and y and then runs through them together.

Reply:
updated that code again a bit